documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,720 @@
+    
+    // Requires that the length of the deadzones span matches the number of axes in the group
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, ReadOnlySpan<Deadzone> deadzones);
+}
+```
+
+# Outputs
+Many controllers have haptics, LEDs, etc, that would be nice to access in a generic, normalized way as well. The following is a first draft concept for this functionality that could accomplish this while leaving room for different sorts of outputs later on. 
+
+Feedback requested: is custom data handling necessary here? At that point it violates the "axis" concept pretty strongly and is best served by higher-level interfaces, but I wanted to make sure we at least consider these things.
+
+```csharp
+public partial interface IAxisDevice
+{
+    /// <summary>
+    /// Output definitions for this device - commonly haptics, LEDs, etc
+    /// </summary>
+    IReadOnlyList<OutputDescription> Outputs { get; }
+    
+    /// <summary>
+    /// Allows the caller to set an output value for a specific output index.
+    /// Valid for any axis not marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// </summary>
+    public void SetOutput(int index, float value);
+    
+    /// <summary>
+    /// Allows the caller to set an output value(s) for a non-floating-point output.
+    /// Valid only for axes marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// Requires that the caller knows the correct data type for the output, and as such the implementation must throw
+    /// an exception if the data type is not correct.
+    /// <br/><br/>
+    /// More than likely, concrete implementations or higher-level interfaces will provide a safer way to set these
+    /// sorts of values. The absence of higher-level utilities for this method should be considered a missing feature
+    /// if the implementation utilizes <see cref = "OutputAxisTrait.RawDataOnly"/>.
+    ///<br/><br/>
+    /// It is more or less expected that this will have few, if any, implementations in the early stages of this API.
+    /// Its use is reserved for advanced use-cases or higher-level implementations where the caller knows the exact
+    /// data type required the output - for example, setting a specific bit pattern for a specific LED, driving
+    /// embedded displays, built-in speakers, etc.
+    /// </summary>
+    /// <param name="index">The index of the <see cref="OutputDescription"/> in <see cref="Outputs"/></param>
+    /// <param name="values">A pointer to the value(s) to be set</param>
+    public void SetOutput<T>(int index, T* values, int count = 1) where T : unmanaged => throw new NotImplementedException();
+}
+
+public enum OutputAxisTrait : ulong
+{
+    None = 0,
+    LED = 1u << 0,
+    Haptic = 1u << 1,
+    
+    LEDBrightness = 1 << 4 | LED,
+    LEDRed = 1u << 5 | LED,
+    LEDGreen = 1u << 6 | LED,
+    LEDBlue = 1u << 7 | LED,
+    LEDWhite = 1u << 8 | LED,

```

Comment:
Kind of feels like we need output groups as well. Which I guess makes sense from a consistency standpoint.

Reply:
it does feel that way huh. could get rid of all these different sub-types of LED in favor of an LED OutputAxisGroupTypeNamePending

Reply:
Basically I think we should just have a very similar API surface for outputs as we do axes - we worked hard on making that generalised abstraction and I like the trait system that you came up with e.g. if it's an LED and it has 3 axes you can assume they're RGB etc